Add Tesla Fuzzy Fingerprint

ssysm opened the pull request:
Allow Tesla to be fuzzy matched. Using the the `TeM` identifier for make id and model code after `(25),`.

sshane commented:
We need to make sure that this can restrict Teslas within a certain model year/capability range. For example, not HW2 (or others), and not Model S/X right now.

sshane commented on opendbc/car/tesla/values.py:
```diff
@@ -34,6 +34,36 @@ class TeslaCarDocsHW4(CarDocs):
+  return candidates
+
+def decode_tesla_fingerprint_model(fingerprint: bytes) -> tuple:
+  if not fingerprint.startswith(b'TeM'): # Not a tesla
```
```suggestion
  if not fingerprint.startswith(b'TeM'):  # Not a tesla
```

two spaces before comments

sshane commented on opendbc/car/tesla/values.py:
```diff
@@ -34,6 +34,36 @@ class TeslaCarDocsHW4(CarDocs):
+  if len(model_split) != 2:
+    return None
+  model = model_split[1][0]
+  return chr(model)
```
two lines between functions

sshane commented on opendbc/car/tesla/values.py:
```diff
@@ -34,6 +34,36 @@ class TeslaCarDocsHW4(CarDocs):
+  model_split = fingerprint.split(b'),')
+  if len(model_split) != 2:
+    return None
+  model = model_split[1][0]
```
this should be robust to garbage input, i.e. no Python exceptions. We use a simple regex pattern for other brands, it ensures the parsing logic is in one place and if there's no match, it's obvious what it will return (None)

sshane commented:
Also see this comment: https://github.com/commaai/opendbc/issues/1917#issuecomment-2740315033

ssysm commented:
Will check, thanks for the info

sshane commented:
We can potentially detect these missing APIs in the interface and set the dashcam flag if needed: https://github.com/commaai/opendbc/blob/1cc799801f03a45180de1e64193981b102aec5c4/opendbc/car/ford/interface.py#L70-L79

ssysm commented:
Changed the firmware decode to a regex, should be able to extract the model code from the firmware. Added a VIN check to restrict only 2018+ cars [Math](https://www.findmyelectric.com/tesla-vin-decoder/). 

As of HW2.5.. I'm not really sure, https://github.com/commaai/opendbc/issues/1917#issuecomment-2740315033. As far I'm seeing in from the route that lukas posted, the `autoPark` signal stayed at `unavailable` the whole trip in `DI_State` and `DAS_Control`  message disappeared after few seconds. Maybe this can give a clue about if ACC is equipped on HW2.5?

sshane commented on opendbc/car/tesla/values.py:
```diff
@@ -36,34 +37,50 @@ class TeslaPlatformConfig(PlatformConfig):
   candidates: set[str] = set()
+
+  # Check year is greater than 2018, but only if VIN is valid
+  if ord(vin[9]) < 74 and vin != '00000000000000000':
```
You can keep the vin constants

lukasloetkolben commented on opendbc/car/tesla/values.py:
```diff
@@ -110,3 +159,14 @@ class TeslaFlags(IntFlag):
 
 STEER_THRESHOLD = 0.5
+
+AP_VERSIONS = {'HW3':'3', 'HW4':'YG4'}
```
YG4 is not an identifier for HW4

sshane commented:
We should only ever use the VIN or the fingerprint to make this easier to reason about
